File: Kata02/Ruby/chop/lib/chop.rb
require "chop/version"

module Chop
    def Chop.chop(item, array)
        low = 0
        width = array.length
        while width > 0 do
            mid_index = width / 2 + low
            if array[mid_index] > item
            elsif array[mid_index] < item
                low = mid_index + 1
                width -= 1
            else
                return mid_index
            end
            width /= 2
        end
        -1
    end
end

File: Kata02/Scala/chop/src/main/scala/Chop.scala
object Chop {
  def main(args: Array[String]) {
    println("Chop!")
  }

  def chop[T](item : T, seq : Seq[T], indexOffset : Int = 0)(implicit f : T => Ordered[T]) : Int = {
    if(seq.nonEmpty) {
      val midIndex = seq.length / 2
      if(seq.apply(midIndex) > item) {
        chop(item, seq.slice(0, midIndex), indexOffset)
      } else if(seq.apply(midIndex) < item) {
        chop(item, seq.slice(midIndex + 1, seq.length), indexOffset + midIndex + 1)
      } else {
        indexOffset + midIndex
      }
    } else {
      -1
    }
  }
}

